docs/broker/filtering.md

```diff
@@ -0,0 +1,336 @@
+# Advanced Filtering for Triggers Proposal
+
+## Problem
+
+As an `Event Consumer` I want to filter events in triggers with a variety of
+strategies beyond exact match on predefined fields. Filtering strategies that
+have been requested include:
+
+*   Filtering by source or type prefix or regular expression
+*   Filtering by a custom extension field
+*   Filtering by data fields
+
+The existing filter capability of exact match on source and type is simple to
+understand and use, but doesn't support these more advanced filtering scenarios.
+
+## Objective
+
+Design a flexible filtering mechanism that supports all of the advanced use
+cases identified and can also implement simpler cases like exact match.
+
+The filtering mechanism MUST be safe and efficient. It SHOULD be compatible with
+a future
+[upstream propagation solution](https://github.com/knative/eventing/issues/934)
+for delivery efficiency at scale.
+
+Filters should be easy for the user to specify and simple to interpret at a
+glance.
+
+## Requirements
+
+*   Support all of the advanced filtering use cases identified.
+*   Can implement simpler cases like exact match.
+*   Safe and secure when embedded in multi-tenant processes.
+*   Compatible with upstream filter propagation.
+*   Easily specified by the user.
+*   Simple for users to interpret.
+
+## Proposed Solution
+
+Add a field to the Trigger Filter spec allowing users to specify an expression
+as a string. The expression is evaluated for each event considered by the
+Trigger. If the expression evaluates to true, the event is delivered to the
+Trigger's subscriber.
+
+The evaluation environment of the expression may include:
+
+*   Standard CloudEvent fields like type and source
+*   Custom extensions dynamically parsed from the event
+*   Data fields parsed from the event, if the data content type is known to be
+    parseable
+
+Since parsing dynamic extension and data fields may be costly, the Trigger must
+explicitly enable it by setting boolean fields adjacent to the expression. See
+[examples](#examples) for syntax.
+
+The expression language chosen is
+[CEL](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/doc/intro.md)
+as it has the best combination of features for this purpose. See
+[alternatives considered](#alternatives-considered) for comparisons to other
+expression languages.
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" &&
+        source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Why CEL?
+
+CEL is the
+[Common Expression Language](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/doc/langdef.md).
```

Comment:
Why link to a specific commit?

Reply:
This is the version of the language definition that was current when this proposal was proposed. Linking to this specific version ensures that future updates to the linked file (including moves) don't alter or obscure the intent of the proposal.